kingpin/actors/base.py

```diff
@@ -498,6 +499,181 @@ def execute(self):
+
+            @gen.coroutine
+            def _get_description(self):
+                yield self.conn.get_desc_of_resource(
+                    name=self.option('name'))
+    """
+
+    # A list of option names that are _not_ automatically managed. These are
+    # useful if you have special behaviors like 'commit' on change, or if you
+    # have parameters that are unmutable ('name').
+    unmanaged_options = []
+
+    def __init__(self, *args, **kwargs):
+        # The 'state' parameter is a given, so make sure its set,
+        self.all_options['state'] = (
+            STATE, 'present', 'Desired state: present or absent')
+
+        # Now go ahead and validate all of the user inputs the normal way
+        super(EnsurableBaseActor, self).__init__(*args, **kwargs)
+
+        # Finally, do a class validation... make sure that we have actual
+        # getter/setter methods for each of the options. This populates dicts
+        # that provide references to the actual methods for execution later.
+        self._verify_methods()
+
+    def _verify_methods(self):
+        """Generates pointers to the Getter and Setter methods.
+
+        Walks through the list of options in self.all_options and discovers the
+        pointers to the getter/setter methods. If any are missing, throws an
+        exception quickly.
+        """
+        self.setters = {}
+        self.getters = {}
+        for option in self.all_options:
+            setter = '_set_%s' % option
+            getter = '_get_%s' % option
+
+            if not self._is_method(getter) or not self._is_method(setter):
+                if option in self.unmanaged_options:
+                    continue
+
+                raise exceptions.UnrecoverableActorFailure(
+                    'Invalid Actor Code Detected in %s: '
+                    'Unable to find required methods: %s, %s'
+                    % (self.__class__.__name__, setter, getter))
+
+            self.setters[option] = getattr(self, setter)
+            self.getters[option] = getattr(self, getter)
+
+    def _is_method(self, name):
+        return hasattr(self, name) and inspect.ismethod(getattr(self, name))
+
+    @gen.coroutine
+    def _ensure(self, option):
+        """Compares the desired state with the actual state of a resource.
+
+        Uses the getter for a resource option to determine its current state,
+        and then compares it with the desired state. Generally does a simple
+        string comparison of the states, but user can optionally define their
+        own comparison mechanism as well.
+
+        If the states do not match, then the setter method is called.
+        """
+        existing = yield self.getters[option]()
+        new = self.option(option)
+
+        # If there is a custom comparison method, use that.. otherwise fall
+        # back to a simple comparison
+        if self._is_method('_compare_%s' % option):
+            equals = yield getattr(self, '_compare_%s' % option)()
```

Comment:
if the getter is a heavy operation then it's done for nothing here. Grab the existing and the new values only in the else block.

Alternatively you can dynamically create the `_compare_{option}` method inside of your `_verify_methods` such that it defaults to string comparison unless it's overwritten by a child class